src/config.rs

```diff
@@ -140,20 +148,40 @@ impl Config {
             );
             self.recovery_threads = MIN_RECOVERY_THREADS;
         }
-        if !Version::is_valid(self.format_version) {
-            warn!(
-                "format-version ({}) is invalid, setting it to {}",
-                self.format_version,
-                Version::default() as u64
-            );
-            self.format_version = Version::default() as u64;
+        if self.enable_log_recycle {
+            if !self.format_version.has_log_signing() {
+                return Err(box_err!(
+                    "format_version: {:?} is invalid when 'enable_log_recycle' on, setting it to V2",
+                    self.format_version
+                ));
+            }
+            if self.purge_threshold.0 / self.target_file_size.0 >= std::u32::MAX as u64 {
+                return Err(box_err!(
+                    "File count exceed UINT32_MAX, calculated by 'purge-threshold / target-file-size'"
+                ));
+            }
         }
         #[cfg(not(feature = "swap"))]
         if self.memory_limit.is_some() {
             warn!("memory-limit will be ignored because swap feature is not enabled");
         }
         Ok(())
     }
+
+    /// Returns the capacity for recycling log files.
+    pub fn recycle_capacity(&self) -> usize {
+        // Attention please, log files with Version::V1 could not be recycled, it might
+        // cause LogBatchs in a mess in the recycled file, where the reader might get
+        // an obsolete entries (unexpected) from the recycled file.
+        if !self.format_version.has_log_signing() {
+            return 0;
+        }
+        if self.enable_log_recycle && self.purge_threshold.0 >= self.target_file_size.0 {
+            (self.purge_threshold.0 / self.target_file_size.0) as usize
```

Comment:
Shall we also restrict the total file number of cache files? In the current logic, the total log size will like to be near `self.purge_threshold` even if there is no much active write traffic

Reply:
No, Raft Engine size always peaks at purge_threshold, even if the load is low. Restricting the stale files will hurt performance.